pkg/skaffold/deploy/kpt.go

```diff
@@ -80,12 +96,79 @@ func (k *KptDeployer) Cleanup(ctx context.Context, _ io.Writer) error {
 	return nil
 }
 
-func (k *KptDeployer) Render(ctx context.Context, out io.Writer, builds []build.Artifact, offline bool, filepath string) error {
-	return nil
+func (k *KptDeployer) Render(ctx context.Context, out io.Writer, builds []build.Artifact, _ bool, filepath string) error {
+	manifests, err := k.renderManifests(ctx, out, builds)
+	if err != nil {
+		return err
+	}
+
+	return outputRenderedManifests(manifests.String(), filepath, out)
+}
+
+func (k *KptDeployer) renderManifests(ctx context.Context, out io.Writer, builds []build.Artifact) (deploy.ManifestList, error) {
+	debugHelpersRegistry, err := config.GetDebugHelpersRegistry(k.globalConfig)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving debug helpers registry: %w", err)
+	}
+
+	manifests, err := k.runKpt(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("running kpt functions: %w", err)
+	}
+
+	if len(manifests) == 0 {
+		return nil, nil
+	}
+
+	manifests, err = manifests.ReplaceImages(builds)
+	if err != nil {
+		return nil, fmt.Errorf("replacing images in manifests: %w", err)
+	}
+
+	for _, transform := range manifestTransforms {
+		manifests, err = transform(manifests, builds, Registries{k.insecureRegistries, debugHelpersRegistry})
+		if err != nil {
+			return nil, fmt.Errorf("unable to transform manifests: %w", err)
+		}
+	}
+
+	return manifests.SetLabels(k.labels)
+}
+
+// runKpt does a dry run with the specified kpt functions (fn-path XOR image) against dir.
+// If neither fn-path or image are specified, functions will attempt to be discovered in dir.
+// An error is returned when both fn-path AND image are specified.
+func (k *KptDeployer) runKpt(ctx context.Context) (deploy.ManifestList, error) {
+	var manifests deploy.ManifestList
+
+	flags := []string{"--dry-run"}
+
+	if len(k.Fn.FnPath) > 0 {
```

Comment:
nit: improve the readability. 

count := 0
if len(k.Fn.FnPath) > 0 {
	count +=1
	flags = append(flags, "--fn-path", k.Fn.FnPath)
} 
if len(k.Fn.Image) > 0 {
	count +=1
	flags = append(flags, "--image", k.Fn.Image)
} 
if len(count) > 1 {
	return err...
}

Reply:
Adding the count seems a little unnecessary to me. There could just be a third check for if both strings are not empty string
```
if len(k.Fn.FnPath) > 0 {
    flags = append(flags, "--fn-path", k.Fn.FnPath)
}
if len(k.Fn.Image) > 0 {
    flags = append(flags, "--image", k.Fn.Image)
}
if k.Fn.FnPath != "" && k.Fn.Image != "" {
    return err...
}
```

Reply:
ooops, I mean `if count > 1`

@MarlonGamez  This suggestion is more about readability (less about logical correctness, more about showing the intend of an implementation). You can find similar examples in k8s CLI tools 
See [example](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/someteam.example.com/v1/replacementtransformer/ReplacementTransformer.go#L51)